File: 2411-spiral-matrix-iv/2411-spiral-matrix-iv.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def spiralMatrix(self, m: int, n: int, head: Optional[ListNode]) -> List[List[int]]:
        matrix = [[-1 for col in range(n)] for row in range(m)]
        direction = 0
        top = 0
        down = m-1
        left = 0
        right = n-1
        res=[]
        counter=1
        while(top <= down and left <= right):
            if direction == 0:
                for i in range(left,right+1):
                    if head is not None:
                        matrix[top][i]=head.val
                        head = head.next
                    else:
                        return matrix
                direction = (direction + 1)%4
                top+=1
            elif direction == 1:
                for i in range(top,down+1):
                    if head is not None:
                        matrix[i][right]=head.val
                        head = head.next
                    else:
                        return matrix
                direction = (direction + 1)%4
                right-=1
            elif direction == 2:
                for i in range(right,left-1,-1):
                    if head is not None:
                        matrix[down][i]=head.val
                        head = head.next
                    else:
                        return matrix
                direction = (direction + 1)%4
                down-=1
            elif direction == 3:
                for i in range(down,top-1,-1):
                    if head is not None:
                        matrix[i][left]=head.val
                        head = head.next
                    else:
                        return matrix
                direction = (direction + 1)%4
                left+=1
        return matrix
        

File: 2411-spiral-matrix-iv/2411-spiral-matrix-iv.kt
class Solution {
    fun spiralMatrix(m: Int, n: Int, head: ListNode?): Array<IntArray> {
        val result = Array(m) { IntArray(n) { -1 } }

        var direction = Direction.RIGHT
        var x = 0
        var y = 0
        var node = head

        while (node != null) {
            result[y][x] = node.`val`
            y += direction.dy
            x += direction.dx
            if (y < 0 || y >= m || x < 0 || x >= n || result[y][x] >= 0) {
                x -= direction.dx
                y -= direction.dy
                direction = direction.next()
                y += direction.dy
                x += direction.dx
            }
            node = node.next
        }
        return result
    }

    enum class Direction(val dx: Int, val dy: Int) {
        RIGHT(1, 0),
        DOWN(0, 1),
        LEFT(-1, 0),
        UP(0, -1);

        fun next(): Direction = items[(ordinal + 1) and 0x03]

        companion object {
            val items = Direction.values()
        }
    }
}